bpf/lib/overloadable_xdp.h

```diff
@@ -184,35 +159,88 @@ static __always_inline bool ctx_snat_done(struct xdp_md *ctx)
 
 #ifdef HAVE_ENCAP
 static __always_inline __maybe_unused int
-ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
-		   __u32 node_id __maybe_unused,
-		   __u32 seclabel __maybe_unused,
-		   __u32 dstid __maybe_unused,
-		   __u32 vni __maybe_unused,
-		   const void *opt __maybe_unused,
-		   __u32 opt_len __maybe_unused,
-		   bool is_ipv6 __maybe_unused,
-		   int *ifindex __maybe_unused)
-{
-	ctx_store_meta(ctx, CB_ENCAP_NODEID, bpf_ntohl(node_id));
-	ctx_store_meta(ctx, CB_ENCAP_SECLABEL, seclabel);
-	ctx_store_meta(ctx, CB_ENCAP_DSTID, dstid);
-	ctx_set_xfer(ctx, XFER_PKT_ENCAP);
-#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
-	if (opt) {
-		if (!is_ipv6) {
-			const __be32 *addr = opt + sizeof(struct geneve_opt_hdr);
-			const __be16 *port = opt + sizeof(struct geneve_opt_hdr) + sizeof(__be32);
-
-			ctx_store_meta(ctx, CB_ENCAP_PORT, *port);
-			ctx_store_meta(ctx, CB_ENCAP_ADDR, *addr);
-		} else {
-			return DROP_DSR_ENCAP_UNSUPP_PROTO;
+ctx_set_encap_info(struct xdp_md *ctx, __be32 src_ip, __be16 src_port,
+		   __u32 daddr, __u32 seclabel __maybe_unused,
+		   __u32 vni __maybe_unused, void *opt, __u32 opt_len, int *ifindex)
+{
+	__u32 inner_len = ctx_full_len(ctx);
+	__u32 tunnel_hdr_len = 8; /* geneve / vxlan */
+	void *data, *data_end;
+	struct ethhdr *eth;
+	struct udphdr *udp;
+	struct iphdr *ip4;
+	__u32 outer_len;
+
+	/* Add space in front (50 bytes + options) */
+	outer_len = sizeof(*eth) + sizeof(*ip4) + sizeof(*udp) + tunnel_hdr_len + opt_len;
+
+	if (ctx_adjust_hroom(ctx, outer_len, BPF_ADJ_ROOM_NET, ctx_adjust_hroom_flags()))
+		return DROP_INVALID;
+
+	/* validate access to outer headers: */
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+
+	if (data + outer_len > data_end)
+		return DROP_INVALID;
+
+	eth = data;
+	ip4 = (void *)eth + sizeof(*eth);
+	udp = (void *)ip4 + sizeof(*ip4);
+
+	memset(data, 0, sizeof(*eth) + sizeof(*ip4) + sizeof(*udp) + tunnel_hdr_len);
+
+	switch (TUNNEL_PROTOCOL) {
+	case TUNNEL_PROTOCOL_GENEVE:
+		{
+			struct genevehdr *geneve = (void *)udp + sizeof(*udp);
+
+			if (opt_len > 0)
+				memcpy((void *)geneve + sizeof(*geneve), opt, opt_len);
+
+			geneve->opt_len = (__u8)(opt_len >> 2);
```

Comment:
I think we can set the critical bit in the Geneve header, right?

Here is the pcap of a Geneve packet 
<img width="657" alt="geneve-xdp" src="https://github.com/cilium/cilium/assets/5572688/971fb5fc-d0bd-4210-abb1-bf53f7f8a2d1">

Reply:
If I remember correctly, the conclusion from https://github.com/cilium/cilium/pull/23890 was that we currently can't set the critical bit from the TC path - correct? Then let's discuss this aspect in a follow-on PR - imho it's not clear whether we want different behaviour from XDP and TC tunneling.

Reply:
Correct. There's no kernel API to set the critical bit for now. I thought we could set it by ourselves here, but year, let's discuss it in a follow-on.

Reply:
I've parked a WIP [here](https://github.com/cilium/cilium/pull/25959), so we don't lose track of it